packages/styled-components/src/models/StyledNativeComponent.js

```diff
@@ -1,152 +1,148 @@
+    for (key in resolvedAttrDef) {
+      context[key] = resolvedAttrs[key] = resolvedAttrDef[key];
+    }
+    /* eslint-enable guard-for-in */
+  });
 
-    if (!attrs.length) return context;
+  return [context, resolvedAttrs];
+}
 
-    this.attrs = {};
+interface StyledComponentWrapperProperties {
+  attrs: Attrs;
+  inlineStyle: Function;
+  displayName: string;
+  foldedComponentIds: Array<string>;
+  target: Target;
+  shouldForwardProp: ?(prop: string, isValidAttr: (prop: string) => boolean) => boolean;
+  styledComponentId: string;
+}
 
-    attrs.forEach(attrDef => {
-      let resolvedAttrDef = attrDef;
-      let attr;
-      let key;
+type StyledComponentWrapper<Config, Instance> = AbstractComponent<Config, Instance> &
+  StyledComponentWrapperProperties;
 
-      if (isFunction(resolvedAttrDef)) {
-        resolvedAttrDef = resolvedAttrDef(context);
-      }
+// Validator defaults to true if not in HTML/DOM env
+const validAttr = () => true;
 
-      /* eslint-disable guard-for-in */
-      for (key in resolvedAttrDef) {
-        attr = resolvedAttrDef[key];
-        this.attrs[key] = attr;
-        context[key] = attr;
-      }
-      /* eslint-enable */
-    });
+function useStyledComponentImpl<Config: {}, Instance>(
+  forwardedComponent: StyledComponentWrapper<Config, Instance>,
+  props: Object,
+  forwardedRef: Ref<any>
+) {
+  const {
+    attrs: componentAttrs,
+    inlineStyle,
+    // $FlowFixMe
+    defaultProps,
+    // $FlowFixMe
+    shouldForwardProp,
+    styledComponentId,
+    target,
+  } = forwardedComponent;
 
-    return context;
-  }
+  useDebugValue(styledComponentId);
 
-  generateAndInjectStyles(theme: any, props: any) {
-    const { inlineStyle } = props.forwardedComponent;
+  // NOTE: the non-hooks version only subscribes to this when !componentStyle.isStatic,
+  // but that'd be against the rules-of-hooks. We could be naughty and do it anyway as it
+  // should be an immutable value, but behave for now.
+  const theme = determineTheme(props, useContext(ThemeContext), defaultProps);
 
-    const executionContext = this.buildExecutionContext(
-      theme,
-      props,
-      props.forwardedComponent.attrs
-    );
+  const [context, attrs] = useResolvedAttrs(theme || EMPTY_OBJECT, props, componentAttrs);
 
-    return inlineStyle.generateStyleObject(executionContext);
-  }
+  const generatedStyles = inlineStyle.generateStyleObject(context);
 
-  setNativeProps(nativeProps: Object) {
-    if (this.root !== undefined) {
-      // $FlowFixMe
-      this.root.setNativeProps(nativeProps);
-    } else if (process.env.NODE_ENV !== 'production') {
-      // eslint-disable-next-line no-console
-      console.warn(
-        'setNativeProps was called on a Styled Component wrapping a stateless functional component.'
-      );
+  const refToForward = forwardedRef;
+
+  const elementToBeCreated: Target = attrs.$as || props.$as || attrs.as || props.as || target;
+
+  const isTargetTag = isTag(elementToBeCreated);
```

Comment:
I know we had this in the codebase before, but I believe that was and is a mistake, since string tags aren't a thing in React Native, so we can assume `isTargetTag === false`